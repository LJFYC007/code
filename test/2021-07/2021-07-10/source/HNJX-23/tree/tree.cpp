#include<bits/stdc++.h>

using namespace std;

#define REP(i, a, b) for ( int i = (a), _end_ = (b); i <= _end_; ++ i )
#define pii pair<int, int>
#define lson root << 1
#define rson root << 1 | 1

const int maxn = 4e5 + 10;

int n, m, Begin[maxn], Next[maxn], To[maxn], e, f[maxn], g[maxn], h[maxn];
int dis[maxn], anc[maxn], Root = 1, size[maxn], id[maxn], dfn[maxn], top[maxn], tot, son[maxn]; 
struct node { int f, g, a, sf, sg, sa, lazy, tag; } Tree[maxn << 2];

inline void PushUp(int root) 
{
    Tree[root].a = Tree[lson].a ^ Tree[rson].a;
    Tree[root].sf = Tree[lson].sf ^ Tree[rson].sf;
    Tree[root].sg = Tree[lson].sg ^ Tree[rson].sg;
    Tree[root].sa = Tree[lson].sa ^ Tree[rson].sa;
}

inline void PushTag1(int root, int lazy)
{
    Tree[root].lazy = lazy; 
    Tree[root].a = lazy == 1 ? Tree[root].f : Tree[root].g;
    Tree[root].sa = lazy == 1 ? Tree[root].sf : Tree[root].sg;
}

inline void PushTag2(int root)
{
    Tree[root].tag ^= 1;
    Tree[root].sa ^= Tree[root].a; Tree[root].sf ^= Tree[root].f; Tree[root].sg ^= Tree[root].g;
}

inline void PushDown(int root)
{
    if ( Tree[root].lazy ) { PushTag1(lson, Tree[root].lazy); PushTag1(rson, Tree[root].lazy); Tree[root].lazy = 0; }
    if ( Tree[root].tag ) { PushTag2(lson); PushTag2(rson); Tree[root].tag = 0; }
}

inline void Build(int root, int l, int r)
{
    if ( l == r ) 
    {
        Tree[root].f = Tree[root].sf = Tree[root].a = Tree[root].sa = f[id[l]];
        Tree[root].g = Tree[root].sg = g[son[id[l]]]; return ;
    }
    int Mid = l + r >> 1; Build(lson, l, Mid); Build(rson, Mid + 1, r);
    Tree[root].f = Tree[lson].f ^ Tree[rson].f; Tree[root].g = Tree[lson].g ^ Tree[rson].g;
    PushUp(root);
}

inline void Modify1(int root, int l, int r, int L, int R)
{
    if ( L <= l && r <= R ) { PushTag2(root); return ; }
    int Mid = l + r >> 1; PushDown(root);
    if ( L <= Mid ) Modify1(lson, l, Mid, L, R);
    if ( Mid < R ) Modify1(rson, Mid + 1, r, L, R);
    PushUp(root);
}

inline void Modify2(int root, int l, int r, int pos, int val)
{
    if ( l == r ) 
    { 
        Tree[root].a = val;
        Tree[root].sa = Tree[root].tag ? 0 : val;
        return ; 
    }
    int Mid = l + r >> 1; PushDown(root);
    if ( pos <= Mid ) Modify2(lson, l, Mid, pos, val);
    else  Modify2(rson, Mid + 1, r, pos, val);
    PushUp(root);
}

inline void Modify3(int root, int l, int r, int L, int R, int val)
{
    if ( L > R ) return ;
    if ( L <= l && r <= R ) { PushTag1(root, val); return ; }
    int Mid = l + r >> 1; PushDown(root);
    if ( L <= Mid ) Modify3(lson, l, Mid, L, R, val);
    if ( Mid < R ) Modify3(rson, Mid + 1, r, L, R, val);
    PushUp(root);
}

inline void add(int u, int v) { To[++ e] = v; Next[e] = Begin[u]; Begin[u] = e; }

inline void DFS1(int u, int fa)
{
    size[u] = 1;
    for ( int i = Begin[u]; i; i = Next[i] ) 
    {
        int v = To[i]; if ( v == fa ) continue ; 
        dis[v] = dis[u] + 1; anc[v] = u;
        DFS1(v, u); size[u] += size[v];
        if ( size[v] > size[son[u]] ) son[u] = v;
        f[u] = max(f[u], f[v] + 1);
    }
}

inline void DFS2(int u, int fa)
{
    int Max = -1, pos = 0, Maxx = -1;
    for ( int i = Begin[u]; i; i = Next[i] ) 
    {
        int v = To[i]; if ( v == fa ) continue ;
        if ( f[v] > Max ) { Maxx = Max; Max = f[v]; pos = v; }
        else Maxx = max(Maxx, f[v]);
    }
    for ( int i = Begin[u]; i; i = Next[i] ) 
    {
        int v = To[i]; if ( v == fa ) continue ; 
        if ( v == pos ) g[v] = max(g[u], Maxx) + 1;
        else g[v] = max(g[u], Max) + 1;
        h[v] = g[v] + 1; DFS2(v, u);
    }
}

inline void DFS3(int u, int t)
{
    dfn[u] = ++ tot; id[tot] = u; top[u] = t;
    if ( son[u] ) DFS3(son[u], t);
    for ( int i = Begin[u]; i; i = Next[i] ) 
    {
        int v = To[i]; if ( v == son[u] || v == anc[u] ) continue ;
        DFS3(v, v);
    }
}

inline void Change(int l, int r) { if ( l <= r ) Modify1(1, 1, n, l, r); }
inline int LCA(int x, int y)
{
    while ( top[x] != top[y] ) { if ( dis[top[x]] < dis[top[y]] ) swap(x, y); x = anc[top[x]]; }
    return dis[x] < dis[y] ? x : y;
}

inline void Rotate(int x)
{
    Modify2(1, 1, n, dfn[x], max(h[x], f[x])); 
    int t = LCA(Root, x);
    while ( top[Root] != top[t] ) 
    {
        Modify3(1, 1, n, dfn[top[Root]], dfn[Root], 1);
        Root = anc[top[Root]];
    }
    Modify3(1, 1, n, dfn[t] + 1, dfn[Root], 1);
    Root = x;
    while ( top[x] != top[t] ) 
    {
        Modify3(1, 1, n, dfn[top[x]], dfn[x] - 1, 2);
        Modify2(1, 1, n, dfn[anc[top[x]]], g[top[x]]);
        x = anc[top[x]];
    }
    Modify3(1, 1, n, dfn[t], dfn[x] - 1, 2);
}

int main()
{
    freopen("tree.in", "r", stdin);
    freopen("tree.out", "w", stdout);
    scanf("%d%d", &n, &m); g[1] = -1;
    REP(i, 1, n - 1) { int u, v; scanf("%d%d", &u, &v); add(u, v); add(v, u); }
    dis[1] = 1; DFS1(1, 0); DFS2(1, 0); DFS3(1, 1);
    Build(1, 1, n);
    while ( m -- ) 
    {
        int op, u, v, x; scanf("%d", &op);
        if ( op == 1 ) 
        {
            scanf("%d%d%d", &u, &v, &x);
            while ( top[u] != top[v] ) 
            {
                if ( dis[top[u]] < dis[top[v]] ) swap(u, v);
                Change(dfn[top[u]], dfn[u]); u = anc[top[u]];
            }
            if ( dis[u] > dis[v] ) swap(u, v);
            Change(dfn[u], dfn[v]);
        }
        else 
        {
            scanf("%d%d", &u, &x);
            if ( dfn[Root] < dfn[u] || dfn[Root] > dfn[u] + size[u] - 1 ) 
                Change(dfn[u], dfn[u] + size[u] - 1);
            else 
            {
                Change(1, n);
                if ( u != Root ) 
                {
                    int x = Root; while ( dis[anc[top[x]]] > dis[u] ) x = anc[top[x]];
                    if ( anc[top[x]] != u ) x = son[u]; else x = top[x];
                    Change(dfn[x], dfn[x] + size[x] - 1);
                }
            }
        }
        Rotate(x); printf("%d\n", Tree[1].sa); 
    }
    return 0;
}