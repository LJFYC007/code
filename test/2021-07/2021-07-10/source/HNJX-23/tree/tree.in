50 50
1 40
2 9
3 8
5 34
7 26
8 38
9 10
12 22
13 22
16 25
19 22
21 32
22 45
26 4
27 11
28 30
29 31
30 33
32 36
34 31
31 10
10 38
35 25
25 24
36 37
38 50
39 24
40 24
24 14
14 33
33 41
42 4
4 43
43 50
45 37
46 20
20 37
37 49
47 41
41 11
11 23
23 15
15 44
44 17
17 18
18 6
6 48
49 48
48 50
1 31 49 22
1 10 4 47
1 29 26 23
1 11 41 14
1 24 33 48
1 17 24 49
1 33 46 14
1 38 18 6
1 25 15 28
1 3 9 28
1 13 41 2
1 5 28 34
1 47 39 39
1 41 50 15
1 17 24 49
1 34 49 32
1 1 26 6
1 16 11 17
1 10 48 5
1 2 43 9
1 43 36 17
1 40 27 30
1 45 24 18
1 7 17 40
1 6 35 15
1 26 13 29
1 20 28 16
1 47 27 48
1 39 23 22
1 1 16 43
1 10 47 23
1 9 1 21
1 1 8 14
1 34 14 21
1 19 24 15
1 41 43 49
1 37 47 33
1 41 21 25
1 12 14 16
1 10 13 44
1 36 50 18
1 3 44 35
1 21 39 26
1 50 46 21
1 29 37 22
1 4 5 28
1 24 34 33
1 25 46 30
1 10 31 34
1 34 36 32
