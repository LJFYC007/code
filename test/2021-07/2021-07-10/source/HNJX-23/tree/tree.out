1
3
12
23
10
0
23
12
7
0
8
8
11
11
9
5
14
1
6
5
0
15
1
28
3
4
9
1
22
2
5
3
6
1
1
2
0
25
31
4
1
0
16
20
26
0
1
23
23
22
